gax/src/main/java/com/google/api/gax/batching/FlowController.java

```diff
@@ -213,9 +283,65 @@ public void release(long elements, long bytes) {
       outstandingElementCount.release(elements);
     }
     if (outstandingByteCount != null) {
-      // Need to return at most as much bytes as it can be drawn.
-      long permitsToReturn = Math.min(bytes, maxOutstandingRequestBytes);
-      outstandingByteCount.release(permitsToReturn);
+      outstandingByteCount.release(bytes);
+    }
+  }
+
+  /**
+   * Increase flow control limits to allow extra elementSteps elements and byteSteps request bytes
+   * before enforcing flow control.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public synchronized void increaseThresholds(long elementSteps, long byteSteps) {
+    Preconditions.checkArgument(elementSteps >= 0);
+    Preconditions.checkArgument(byteSteps >= 0);
+    setThresholds(
+        currentOutstandingElementCount == null ? 0 : currentOutstandingElementCount + elementSteps,
+        currentOutstandingRequestBytes == null ? 0 : currentOutstandingRequestBytes + byteSteps);
+  }
+
+  /**
+   * Decrease flow control limits to allow elementSteps fewer elements and byteSteps fewer request
+   * bytes before enforcing flow control.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public synchronized void decreaseThresholds(long elementSteps, long byteSteps) {
+    Preconditions.checkArgument(elementSteps >= 0);
+    Preconditions.checkArgument(byteSteps >= 0);
+    setThresholds(
+        currentOutstandingElementCount == null ? 0 : currentOutstandingElementCount - elementSteps,
+        currentOutstandingRequestBytes == null ? 0 : currentOutstandingRequestBytes - byteSteps);
+  }
+
+  /**
+   * Set flow control limits to elements and bytes. If elements or bytes are greater than the
+   * absolute maximum, limits are set to the maximum. If elements or bytes are smaller than the
+   * absolute minimum, limits are set to the minimum.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public synchronized void setThresholds(long elements, long bytes) {
```

Comment:
Should this be private?

Reply:
Consider changing this be `adjustThresholds` and have it take relative amounts. So that you are limiting the number of places that you are touching synchronized state